### ISSUE TITLE
Extend SpamCheck API?

### ISSUE THREAD

**Issue by Yoric** [matrix-org/synapse#10311]
Recent experience suggests that we would benefit from having some more details about messages (and invite requests, etc.) to determine whether a request is spam.

- We have the IP of the user during registrations. But if we only detected that the IP is used for spam *after* registration, we can't block them through the antispam. Ideally, the antispam should have access to both registration IP and current IP, assuming that the user is on the current antispam.
- Some antispams may wish to apply policies such as "can't invite if you have just registered", etc. So, having access to information on when the user has registered would be useful.
- Similarly, how many messages they have already sent that weren't picked as spam.

**Comment by Yoric** [comment]
Ideally, an API to get statistics on users would be great!

**Comment by Yoric** [comment]
Also, perhaps an API to kick/ban/shadowban users who have been identified as spammers?

**Comment by babolivier** [comment]
> Also, perhaps an API to kick/ban/shadowban users who have been identified as spammers?

What would this look like? From Synapse's point of view, what does a user who has "been identified as spammer" look like? Also worth noting that the module API allows module to [send events](https://github.com/matrix-org/synapse/blob/10671da05bdb72d98aab2a8937da503abfc836fd/synapse/module_api/__init__.py#L387) into rooms, so in theory a module can already do this (except for the shadowban bit).

**Comment by Yoric** [comment]
> > Also, perhaps an API to kick/ban/shadowban users who have been identified as spammers?
> 
> What would this look like? From Synapse's point of view, what does a user who has "been identified as spammer" look like?

I mean that if a SpamCheck implementation decides that a user is a spammer, it might make sense to allow that implementation to shadowban/etc. the user.

**Comment by babolivier** [comment]
> if a SpamCheck implementation decides that a user is a spammer

Right, my point is that currently this doesn't mean anything from the Synapse perspective.

**Comment by Yoric** [comment]
> > if a SpamCheck implementation decides that a user is a spammer
> 
> Right, my point is that currently this doesn't mean anything from the Synapse perspective.

Sure, and it doesn't have to mean anything for Synapse, only for the SpamCheck implementation.

**Comment by babolivier** [comment]
Sorry, I'm having a bit of trouble figuring out what exactly you're asking for on this point. Would it be the ability to mark users as shadowbanned from the module API?

**Comment by Yoric** [comment]
> Would it be the ability to mark users as shadowbanned from the module API?

Exactly, in addition to the ability to request IPs.

**Comment by babolivier** [comment]
Got it, thanks for bearing with me on this :+1:
